pyuvdata/uvdata.py

```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+            temp_flag = np.zeros((temp_Nblts, self.Nspws, self.Nfreqs, self.Npols),
+                                 dtype=self.flag_array.dtype)
+            temp_nsample = np.zeros((temp_Nblts, self.Nspws, self.Nfreqs, self.Npols),
+                                    dtype=self.nsample_array.dtype)
+
+        temp_idx = 0
+        for bl in bls_to_downsample:
+            bl_inds = np.nonzero(self.baseline_array == bl)[0]
+            running_int_time = 0.0
+            summing_idx = 0
+            n_sum = 0
+            for itime, int_time in enumerate(self.integration_time[bl_inds]):
+                running_int_time += int_time
+                n_sum += 1
+                over_min_int_time = (running_int_time > min_int_time
+                                     or np.isclose(running_int_time, min_int_time,
+                                                   rtol=self._integration_time.tols[0],
+                                                   atol=self._integration_time.tols[1]))
+                last_sample = (itime == len(bl_inds) - 1)
+                # We sum up all the samples found so far if we're over the target minimum
+                # time, or we've hit the end of the time samples for this baseline.
+                if over_min_int_time or last_sample:
+                    if last_sample and not (over_min_int_time or keep_ragged):
+                        # don't do anything -- implicitly drop these integrations
+                        continue
+                    # sum together that number of samples
+                    temp_baseline[temp_idx] = bl
+                    # this might be wrong if some of the constituent times are *totally* flagged
+                    averaging_idx = bl_inds[summing_idx:summing_idx + n_sum]
+                    temp_time[temp_idx] = np.average(self.time_array[averaging_idx])
+                    temp_int_time[temp_idx] = running_int_time
+                    if not self.metadata_only:
+                        # if all inputs are flagged, the flag array should be True,
+                        # otherwise it should be False.
+                        # The sum below will be zero if it's all flagged and
+                        # greater than zero otherwise
+                        # Then we use a test against 0 to turn it into a Boolean
+                        temp_flag[temp_idx] = np.sum(~self.flag_array[averaging_idx], axis=0) == 0
+
+                        mask = self.flag_array[averaging_idx]
+                        # need to update mask if a downsampled visibility will be flagged
+                        # so that we don't set it to zero
+                        if (temp_flag[temp_idx]).any():
+                            ax1_inds, ax2_inds, ax3_inds = np.nonzero(temp_flag[temp_idx])
+                            for ii in range(ax1_inds.size):
+                                mask[:, ax1_inds[ii], ax2_inds[ii], ax3_inds[ii]] = False
```

Comment:
I'm a little confused at the need to iterate through the arrays here. Is it not able to just be indexed as `mask[:, ax1_inds, ax2_inds, ax3_inds]`?  We're saying here that all indices for the 0th axis should be flagged right?

Reply:
I could be wrong, but I don't think that's quite the same. For example, if `ax1_inds = [0, 1, 2]`, `ax2_inds = [0, 1, 2]`, and `ax3_inds = [0, 1, 2]`, the code as written will set the diagonal to `False`, whereas your code would set the whole thing to `False`. In other words, code as written allows for fine grained selection of each individual pixel, but yours will only unflag slices.

Reply:
we should test this. If we don't need a for loop i would prefer to avoid it.

Reply:
I've tested this empirically. Dropping in the `ax?_inds` seems to Do The Right Thing and just zip them together (rather than take the outer product). I've opted for @mkolopanis's approach to eliminate the for loop.